cmd/build/v2/hasher.go

```diff
@@ -0,0 +1,138 @@
+// Copyright 2023 The Okteto Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v2
+
+import (
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+
+	oktetoLog "github.com/okteto/okteto/pkg/log"
+	"github.com/okteto/okteto/pkg/model"
+)
+
+const (
+	OktetoSmartBuildUsingContextEnvVar = "OKTETO_SMART_BUILDS_USING_BUILD_CONTEXT"
+
+	buildContextCommitType = "tree_hash"
+	projectCommitType      = "commit"
+)
+
+type repositoryCommitRetriever interface {
+	GetSHA() (string, error)
+	GetTreeHash(string) (string, error)
+}
+
+type serviceHasher struct {
+	gitRepoCtrl repositoryCommitRetriever
+
+	projectCommit string
+}
+
+func newServiceHasher(gitRepoCtrl repositoryCommitRetriever) *serviceHasher {
+	return &serviceHasher{
+		gitRepoCtrl: gitRepoCtrl,
+	}
+}
+
+// HashProjectCommit hashes the
+func (sh *serviceHasher) HashProjectCommit(buildInfo *model.BuildInfo) string {
+	if sh.projectCommit == "" {
+		var err error
+		sh.projectCommit, err = sh.gitRepoCtrl.GetSHA()
+		if err != nil {
+			oktetoLog.Infof("could not get repository sha: %w", err)
+		}
+	}
+	return sh.hash(buildInfo, projectCommitType, sh.projectCommit)
+}
+
+func (sh serviceHasher) HashBuildContext(buildInfo *model.BuildInfo) string {
+	buildContext := buildInfo.Context
+	if buildContext == "" {
+		buildContext = "."
+	}
+	treeHash, err := sh.gitRepoCtrl.GetTreeHash(buildContext)
```

Comment:
worth create a singleton also for treeHash? that way we don't calculate it each time we need that value